modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ButtonTag.java

```diff
@@ -254,6 +254,15 @@ protected void cleanUp() {
 		_type = "button";
 	}
 
+	@Override
+	protected String getHydratedModuleName() {
+		if (Validator.isNotNull(getAdditionalProps())) {
+			return "@clayui/button/lib/index";
```

Comment:
I think this would be ideal, but I'm not sure we can pull it off... other than the `renderable` issue commented above, there are issues with hydration we'd need to figure out.

The key here, though is that now that we've coupled the hydration algorithm with the hydratedModuleName, there's no way to generically say hydrate when props without coding it like this in any tag that should support it (which will likely be all).

Reply:
> there's no way to generically say hydrate when props without coding it like this in any tag that should support it (which will likely be all of them).

Do you see that as much of a deal-breaker? These taglibs are write-once, use-many-times. It seems like a fair trade-off to me that *you* have to write this out in a number of taglibs, in order for *countless* devs to enjoy using them (and without ever having to look inside them).

![dozens-of-us](https://user-images.githubusercontent.com/7074/84870659-e060d900-b07f-11ea-8f73-182949335c0a.gif)

(I might also be misunderstanding your concern.)

Reply:
No deal breaker, just pointing it out. That we now avoid having to implement two separate methods for hydration to work, but we will always need to account for `props` in every one of them now. Feels like pushing it from one side to the other, just that, no big deal :)

Reply:
I was arguing against having two separate methods not because of the amount of work involved, but [because the API was not "atomic"](https://github.com/wincent/liferay-portal/pull/351#discussion_r439365053) — that is, the type system would allow you to implement it "wrong" because there is no way to statically verify that the combination of the two is valid.

You're right though: every design decision comes with trade-offs. I don't think there are "right" or "wrong" designs, only compromises. I'm happy with pretty much any API you come up with here, as I don't know enough Java to touch it anyway. 😂

Reply:
Nah, it's okay... OOP at its worst 😂

Reply:
We're in agreement here, resolving this conversation.

TL;DR: @wincent was right ;)